### PR TITLE
Require record fields to be simple identifiers

### PR DESCRIPTION
**Pull request by shonfeder**
When a `::` appears in a quint identifiers, I am referring to that identifier as
"qualified". E.g., `Foo::bar` is qualified, but `_ding` is not.

This takes care of one problem identified in #1081 by introducing a syntax error
a record field is a qualified identifier.

Along the way, the PR also gives further illustration of how we can address #17
by using error grammar rules. (There was already precedent for this in the
`errorEq` rule checking for the misuse of  `=` for equality). This shows how we
can move stronger syntactic checks into the grammar without sacrificing the
quality of our error messages.

E.g., errors for this new rule look like so:

```quint
>>> type r = {a: int, b::c::d: str}
syntax error: error: [QNT008] Identifiers in a record cannot be qualified with '::'. Found b::c::d.
type r = {a: int, b::c::d: str}
                         ^
```

To enable this improvement in the grammar, we had to refactor the lexical rule
for `IDENTIFIER`: without splitting this into a lexical rule for the
unqualified case and a grammar rule for the qualified, we end up with an
ambiguous lexical rule that makes parsing fail.

One incidental change is added to make the `reference` field in our error type
optional. This is implied by the fact that parser errors are errors that should
be represented with this structure, but we may not have a reference for them yet.

- [x] Tests added for any new code
- [-] Documentation added for any new functionality
- [x] Entries added to the respective `CHANGELOG.md` for any new functionality
- [-] Feature table on [`README.md`](../README.md#roadmap) updated for any listed functionality

### CODE REVIEW

**Comment by shonfeder** [comment]
The last two commits are generated/mechanical!

File: quint/src/generated/QuintListener.ts [hunk 1]
```diff
@@ -1,6 +1,10 @@
 // Generated from ./src/generated/Quint.g4 by ANTLR 4.9.0-SNAPSHOT
 
 
+    // Used for forming errors
+    import { quintErrorToString } from '../quintError'
```
Comment (thpani) [inline quint/src/generated/QuintListener.ts:5]: Unusual indent here. Should we dedent it in `Quint.g4`? I'm okay either way.

File: quint/src/generated/Quint.g4 [hunk 2]
```diff
@@ -214,6 +221,21 @@ operator: (AND | OR | IFF | IMPLIES |
...
+// An unqualified identifier that raises an error if a qualId is supplied
+simpleId[context: string]
+    : IDENTIFIER
+    | qualId {
+        const err = quintErrorToString(
+          { code: 'QNT008',
+            message: "Identifiers in a " + $context + " cannot be qualified with '::'. Found " + $qualId.text + "."
+          },
+        )
+        this.notifyErrorListeners(err)
```
Comment (thpani) [inline quint/src/generated/Quint.g4:235]: I'm torn whether we should use `quintErrorToString` here at all (with all it's implications, making error fields optional), or simply pass a string as for `QNT006` above. WDYT?

**Review by thpani — APPROVED** [review]
Looks great!

I have an alternative suggestion to making the error type fields optional; not sure myself if I prefer one over the other.

File: quint/src/generated/Quint.g4 (hunk 2 above)
Comment (shonfeder) [reply quint/src/generated/Quint.g4:235]: So my justification for bringing it in is that we want to put white a few more errors here probably, and using the type and function ensures we are using valid error ids, and if/when we overhaul the error system or add hints to tic etc, we will know to add them here as well. I do agree it’s more overhead, but already in this case it caught a typo on in my error if 😅.

If you feel strongly that we should stick to strings tho I’m ok with that. We could get @bugarela’s opinion as a tiebreaker! :)

regarding making the reference field optional, note that we already effectively treat it as optional (afaict), in that we allow references which aren’t ok our location map and then just print and unknown location.

Comment (thpani) [reply quint/src/generated/Quint.g4:235]: > that we want to put white a few more errors here probably, and using the type and function ensures we are using valid error ids, and if/when we overhaul the error system or add hints to tic etc, we will know to add them here as well

sgtm 👍

File: quint/src/generated/QuintListener.ts (hunk 1 above)
Comment (shonfeder) [reply quint/src/generated/QuintListener.ts:5]: Oh yeah, I didn't look at the generated code! I think removing the indent won't hurt readability of the antlr, so I'll do that.

**Pull request merged**